Fix forward AD add.Scalar to not set tangent as-is

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79767
* __->__ #79664
* #79653

Test plan:
```cpp
b = a.add(1);  // calls into add.Scalar which sets out's tangent to be a's tangent
b.sub_(1); // doing any in-place op will also mutate a's tangent
```

soulitzer committed:
Fix forward AD add.Scalar to not set tangent as-is

[ghstack-poisoned]

soulitzer committed:
Update on "Fix forward AD add.Scalar to not set tangent as-is"


Test plan:
```cpp
b = a.add(1);  // calls into add.Scalar which sets out's tangent to be a's tangent
b.sub_(1); // doing any in-place op will also mutate a's tangent
```

[ghstack-poisoned]

nikitaved commented on tools/autograd/derivatives.yaml:
```diff
@@ -208,7 +208,7 @@
 
 - name: add.Scalar(Tensor self, Scalar other, Scalar alpha=1) -> Tensor
   self: handle_r_to_c(self.scalar_type(), grad)
-  result: self_t
+  result: self_t.clone()
 
```
maybe `auto_element_wise` will do the trick?

soulitzer replied:
I can see float to complex type promotion being something we might want here (I don't know how else people can reach this path without complex literals though, python binding already wraps scalar inputs with tensors):
```cpp
  using namespace c10::complex_literals;

  auto a = torch::ones({3, 3}, c10::TensorOptions().dtype(at::kFloat));
  auto b = a.add(0.5_id);
```
Not sure about `auto_element_wise` since we kind of want the opposite of  `handle_r_to_c` (cast to real) here right?

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -208,7 +208,7 @@
 - name: add.Scalar(Tensor self, Scalar other, Scalar alpha=1) -> Tensor
   self: handle_r_to_c(self.scalar_type(), grad)
-  result: self_t
+  result: self_t.clone()
```
This should be done implicitly already by `set_fw_grad()`. As the "viewness" of the tensor and its tangent don't match.
If that is not happening, I would argue that we should fix `set_fw_grad()`.

albanD replied:
Ho is `self_t` a non-differentiable view here?

soulitzer replied:
Doesn't even have to be. `self_t` can just be a plain tensor here. The issue is that the two dual tensors end up sharing the same tangent when the op is supposed to be out-of-place. Unless set_fw_grad or VariableType checks TensorImpl use count, it would not know this. I think the invariant forward AD should be checking here is: that if the forward is out-of-place, then the forward grad should also be out-of-place (similar to the tensorimpl use count checks we have already).

albanD replied:
From offline discussion: Indeed out of place ops should have out of place formulas.
And since this is a property of the function and not of the Tensor, we do want the function to do this.
We should add a check for these and the `.clone()` is a good solution.

albanD reviewed (approved):
We should add some DEBUG_ONLY asserts to ensure that we don't return one of the input's gradient as-is.
But that can be in a follow up

soulitzer commented:
@pytorchbot merge -g

soulitzer commented:
@pytorchbot rebase

pytorchmergebot committed:
Update on "Fix forward AD add.Scalar to not set tangent as-is"


Test plan:
```cpp
b = a.add(1);  // calls into add.Scalar which sets out's tangent to be a's tangent
b.sub_(1); // doing any in-place op will also mutate a's tangent
```

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g
